### ISSUE TITLE
EndBlock updataValidator

### ISSUE THREAD

**Issue by pj0579** [tendermint/tendermint#1331]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

**Tendermint version** : 0.16.0

**Environment**:
- **OS** mac
- **Install tools**:glide
- **Others**:

**What happened**:
updates to validators                        module=consensus updates="[{\"pub_key\":\"NUI5NEIzRTJDMzRGNkU1NDdBMjY1MUVDNUZCRjc1MjA3Mjk3QzA1ODg4N0NBNUU5Mjk4QzRCRTJDRDg5RkI3Qg==\",\"power\":10}]"
ERROR: Failed to create node: Error starting proxy app connections: Error on replay: Commit failed for application: Error changing validator set: Unexpected type byte 35 for type crypto.PubKeyInner

**What you expected to happen**:
success add validators

**How to reproduce it** (as minimally and precisely as possible):
    @Override
    public ResponseEndBlock requestEndBlock(RequestEndBlock req) {
        System.out.print(req);
        final byte[] validatorPubKey = "5B94B3E2C34F6E547A2651EC5FBF75207297C058887CA5E9298C4BE2CD89FB7B".getBytes();
        ResponseEndBlock.Builder builder = ResponseEndBlock.newBuilder();
        Validator validator =Validator.newBuilder().setPower(10L).setPubKey(ByteString.copyFrom(validatorPubKey)).build();
        builder.addValidatorUpdates(validator);
        return builder.build();
    }

**Comment by ebuchman** [comment]
Can you clarify what happened here? 

Did the validator set successfully change, but then it failed when the node was restarted ?

**Comment by pj0579** [comment]
I just use a node. requestEndBlock function will be called when the node starts and pubkey is constant.but  it failed

**Comment by pj0579** [comment]
error is 
<code>Error changing validator set: Unexpected type byte 35 for type crypto.PubKeyInner</code>

**Comment by srmo** [comment]
This looks like Java, i.e. jABCI. 
@pj0579 I assume that the problem lies in 
`"5B94B3E2C34F6E547A2651EC5FBF75207297C058887CA5E9298C4BE2CD89FB7B".getBytes();` 
The string "5B..." is already the bytes if I'm not mistaken. So calling getBytes() on this string object now returns a byte array which is no longer a valid public key. You need to create a byte array where the first byte is actually "5", then "B" and so on.

**Comment by srmo** [comment]
PS: for example bouncycastle's Hex implementation allows you to do that
`Hex.decode(hex)` or even better the core utility `javax.xml.bind.DatatypeConverter.parseHexBinary(String)`

Sure, you can always use the manual approach :)

**Comment by srmo** [comment]
@ebuchman do you have to pass the bytes of the publicKey or the Base64 representation? Looking at the tendermint log, it shows that it has interpreted the bytes of the publicKey as base64, doesn't it?
`[{"pub_key":"NUI5NEIzRTJDMzRGNkU1NDdBMjY1MUVDNUZCRjc1MjA3Mjk3QzA1ODg4N0NBNUU5Mjk4QzRCRTJDRDg5RkI3Qg==","power":10}]"`

**Comment by ebuchman** [comment]
ABCI expects go-wire encoded pubkeys in the EndBlock msg as raw bytes (not hex or base64 encoded)

**Comment by srmo** [comment]
@ebuchman so this hould be handled by the compiled protobuf types and our correct wire encoding (zigzag and stuff), right?
@pj0579 please check if your code works when using the byte conversion approach I mentioned. If it does not and you are using jABCI, please open an issue at https://github.com/jTendermint/jabci

**Comment by wolfposd** [comment]
Im assuming go-wire needs information about what type of pukey is used?

`5B94B3E2C34F6E547A2651EC5FBF75207297C058887CA5E9298C4BE2CD89FB7B`

this seems to just be the bytes of the public key

**Comment by ebuchman** [comment]
Yes that's right.

Though note for computing the address (a separate concern), we treat the pubkey as `[]byte` rather than `[32]byte`, which means we also prefix the length in the go-wire encoding. So the thing that's hashed is `<0x01><0x0120><pubkey>` where the `<0x01>` is the type byte and the `<0x0120>` is the length (first byte is length of length, second byte is length itself).

I am aware this is confusing. It will change when new go-wire (Amino) is released!

**Comment by srmo** [comment]
Yeah, confusing. The naive assumption was that it was already all encoded by the protobuf types. And having ABCI messages use the new zigzag vs this using the old varint ... Looking forward to the cleanup!

**Comment by zglzglak47** [comment]
I got the same issue for persistent_kvstore when running a 4 node tendermint in docker-compose by setting   `--proxy_app='persistent_kvstore'`

Error log is :
tm_node1      | I[04-17|00:27:16.609] Updates to validators                        module=consensus updates="[{\"pub_key\":\"aSs5ly2s8rSSX/m167HTGpYumTbPQqtfg4JvcWThsLc=\",\"power\":99}]"
tm_node1      | ERROR: Failed to create node: Error starting proxy app connections: Error on replay: Commit failed for application: Error changing validator set: Unexpected type byte 69 for type crypto.PubKeyInner

the genesis file is like :
    ....
    {
      "pub_key": {
        "type": "ed25519",
        "data": "692B39972DACF2B4925FF9B5EBB1D31A962E9936CF42AB5F83826F7164E1B0B7"
      },
      "power": 10,
      "name": ""
    },
....

**Comment by wolfposd** [comment]
> [{"pub_key":"aSs5ly2s8rSSX/m167HTGpYumTbPQqtfg4JvcWThsLc=","power":99}]"
you need to encode the pubkey-type like so:
`<0x01><0x0120><pubkey>`

so resulting key would be:
`010120692B39972DACF2B4925FF9B5EBB1D31A962E9936CF42AB5F83826F7164E1B0B7`

**Comment by zglzglak47** [comment]
I am not sure about `<0x01><0x0120><pubkey>`.  When is use 
`curl http://localhost:46667/broadcast_tx_commit?tx=%22val:010120692B39972DACF2B4925FF9B5EBB1D31A962E9936CF42AB5F83826F7164E1B0B7/50`
to change a validator's power,  the `dump_consensus_state` api shows that I add a new validator with pub key 0120692B39972DACF2B4925FF9B5EBB1D31A962E9936CF42AB5F83826F7164E1.  And then the validators may not able to reach consensus again.

I make a small test for the source pub_key.go. When 
pubkey, err := crypto.PubKeyFromBytes(v.PubKey)
pub key.Bytes() is at length 33.

So I just try `<0x01><pubkey>`, just add 01 before the pubkey in genesis file. It works for me.

**Comment by ebuchman** [comment]
See [current](https://github.com/tendermint/tendermint/blob/develop/docs/specification/new-spec/encoding.md) and [pre amino](https://github.com/tendermint/tendermint/blob/develop/docs/specification/new-spec/pre-amino.md)
